File: src/main/java/groupwork/Customer.java
package groupwork;

import java.util.Enumeration;
import java.util.Vector;

class Customer extends DomainObject
{
    public Customer(String name) {
        _name = name;
    }
    public String statement() {
        double totalAmount = 0;
        int frequentRenterPoints = 0;
        Enumeration rentals = _rentals.elements();
        String result = "Rental Record for " + name() + "\n";
        while (rentals.hasMoreElements()) {
            double thisAmount = 0;
            Rental each = (Rental) rentals.nextElement();

            //determine amounts for each line
            switch (each.tape().movie().priceCode()) {
                case Movie.REGULAR:
                    thisAmount += 2;
                    if (each.daysRented() > 2)
                        thisAmount += (each.daysRented() - 2) * 1.5;
                    break;
                case Movie.NEW_RELEASE:
                    thisAmount += each.daysRented() * 3;
                    break;
                case Movie.CHILDRENS:
                    thisAmount += 1.5;
                    if (each.daysRented() > 3)
                        thisAmount += (each.daysRented() - 3) * 1.5;
                    break;

            }
            totalAmount += thisAmount;

            // add frequent renter points
            frequentRenterPoints ++;
            // add bonus for a two day new release rental
            if ((each.tape().movie().priceCode() == Movie.NEW_RELEASE) && each.daysRented() > 1) frequentRenterPoints ++;

            //show figures for this rental
            result += "\t" + each.tape().movie().name()+ "\t" + String.valueOf(thisAmount) + "\n";

        }
        //add footer lines
        result +=  "Amount owed is " + String.valueOf(totalAmount) + "\n";
        result += "You earned " + String.valueOf(frequentRenterPoints) + " frequent renter points";
        return result;

    }
    public void addRental(Rental arg) {
        _rentals.addElement(arg);
    }
    public static Customer get(String name) {
        return (Customer) Registrar.get("Customers", name);
    }
    public void persist() {
        Registrar.add("Customers", this);
    }
    private Vector _rentals = new Vector();
}

File: src/main/scala/groupwork/refactor/Customer.scala
package groupwork.refactor

import scala.util.Try

case class Customer(name: String, rentals: List[Rental], frequentPoints: Int = 0) extends DomainObject {
  def add(rental: Rental): Customer = copy(rentals = rental :: rentals)
  
  lazy val statement: String = {
    /* todo
    for each rental, calculate individual cost as days rented * cost of rental type
    add a frequent renter point — need to store somewhere
    for each new release rented over one day, add a frequent renter point
    output amount owed and renter points
     */
    
    ??? //todo
  }
}

object Customer {
  private val category = "Customers"
  
  def persist(customer: Customer): Try[Unit] = Registrar.add(category, customer)
  
  def get(name: String): Option[Customer] = Registrar.get(category, name) match {
    case result: Customer => Some(result)
    case _ => None
  }
}
